numpy/core/tests/test_shape_base.py

```diff
@@ -330,5 +333,121 @@ def test_stack():
+        a = np.array([1, 2, 3])
+        b = np.array([2, 3, 4])
+        expected = np.array([[1, 2, 3, 2, 3, 4],
+                             [1, 2, 3, 2, 3, 4]])
+        result = block([[a, b], [a, b]])
+        assert_equal(expected, result)
+
+    def test_block_with_1d_arrays_column_wise(self):
+        # # # 1-D vectors are treated as row arrays
+        a = np.array([1, 2, 3])
+        b = np.array([2, 3, 4])
+        expected = np.array([[1, 2, 3],
+                             [2, 3, 4]])
+        result = block([[a], [b]])
+        assert_equal(expected, result)
+
+    def test_block_mixed_1d_and_2d(self):
+        A = np.ones((2, 2))
+        B = np.array([2, 2])
+        result = block([[A], [B]])
+        expected = np.array([[1, 1],
+                             [1, 1],
+                             [2, 2]])
+        assert_equal(expected, result)
+
+    def test_block_complicated(self):
+        # a bit more complicated
+        One = np.array([[1, 1, 1]])
+        Two = np.array([[2, 2, 2]])
+        Three = np.array([[3, 3, 3, 3, 3, 3]])
+        four = np.array([4, 4, 4, 4, 4, 4])
+        five = np.array([5])
+        six = np.array([6, 6, 6, 6, 6])
+        Zeros = np.zeros((2, 6))
+
+        result = block([[One, Two],
+                        Three,
+                        four,
+                        [five, six],
+                        Zeros])
+        expected = np.array([[1, 1, 1, 2, 2, 2],
+                             [3, 3, 3, 3, 3, 3],
+                             [4, 4, 4, 4, 4, 4],
+                             [5, 6, 6, 6, 6, 6],
+                             [0, 0, 0, 0, 0, 0],
+                             [0, 0, 0, 0, 0, 0]])
+        assert_equal(result, expected)
+
+        # additional [] around rows should not have any influence
+        result = block([[One, Two],
+                        Three,
+                        [four],
+                        [five, six],
+                        Zeros])
+        assert_equal(result, expected)
+
+        result = block([[One, Two],
+                        [Three],
+                        [four],
+                        [five, six],
+                        Zeros])
+        assert_equal(result, expected)
+
+        result = block([[One, Two],
+                        [Three],
+                        [four],
+                        [five, six],
+                        [Zeros]])
+        assert_equal(result, expected)
+
+    def test_block_with_mismatched_shape(self):
+        a = np.array([0, 0])
+        b = np.eye(2)
+        assert_raises(ValueError, np.block, (a, b))
+        assert_raises(ValueError, np.block, (b, a))
+
+    def test_no_lists(self):
+        assert_equal(np.block(1),         np.array(1))
+        assert_equal(np.block(np.eye(3)), np.eye(3))
+
+    def test_empty_input(self):
+        assert_raises(ValueError, np.block, [])
+        assert_raises(ValueError, np.block, [[]])
+
+
```

Comment:
Right. I was thinking of the last item on your enhancements list:
 > Preallocate the result array to save redundant copies